lib/util/Components.js

```diff
@@ -512,43 +512,35 @@ function componentRule(rule, context) {
       let scope = context.getScope();
       while (scope) {
         const node = scope.block;
-        const isFunction = /Function/.test(node.type); // Functions
-        const isArrowFunction = astUtil.isArrowFunction(node);
-        const enclosingScope = isArrowFunction ? utils.getArrowFunctionScope(scope) : scope;
-        const enclosingScopeParent = enclosingScope && enclosingScope.block.parent;
-        const isClass = enclosingScope && astUtil.isClass(enclosingScope.block);
-        const isMethod = enclosingScopeParent && enclosingScopeParent.type === 'MethodDefinition'; // Classes methods
-        const isArgument = node.parent && node.parent.type === 'CallExpression'; // Arguments (callback, etc.)
-        // Attribute Expressions inside JSX Elements (<button onClick={() => props.handleClick()}></button>)
-        const isJSXExpressionContainer = node.parent && node.parent.type === 'JSXExpressionContainer';
-        const pragmaComponentWrapper = this.getPragmaComponentWrapper(node);
-        if (isFunction && pragmaComponentWrapper) {
-          return pragmaComponentWrapper;
-        }
-        // Stop moving up if we reach a class or an argument (like a callback)
-        if (isClass || isArgument) {
-          return null;
-        }
-        // Return the node if it is a function that is not a class method and is not inside a JSX Element
-        if (isFunction && !isMethod && !isJSXExpressionContainer && utils.isReturningJSXOrNull(node)) {
-          return node;
+
+        if (node.type === 'FunctionDeclaration') {
+          if (utils.isReturningJSXOrNull(node)) {
+            return node;
+          }
         }
-        scope = scope.upper;
-      }
-      return null;
-    },
 
-    /**
-     * Get an enclosing scope used to find `this` value by an arrow function
-     * @param {Scope} scope Current scope
-     * @returns {Scope} An enclosing scope used by an arrow function
-     */
-    getArrowFunctionScope(scope) {
-      scope = scope.upper;
-      while (scope) {
-        if (astUtil.isFunction(scope.block) || astUtil.isClass(scope.block)) {
-          return scope;
+        if (node.type === 'FunctionExpression' || node.type === 'ArrowFunctionExpression') {
+          switch (node.parent.type) {
+            case 'VariableDeclarator':
+            case 'AssignmentExpression':
+            case 'Property':
+            case 'ReturnStatement':
+            case 'ExportDefaultDeclaration': {
```

Comment:
These cases are
```js
const Foo = () => <p />;

module.exports = function Foo() {};

module.exports = {
  Foo() {},
};

function hof() {
  return function Foo() {};
}

export default function Foo() {}
```

Reply:
Can we add explicit test cases for each of them?

Reply:
Currently the tests cases for them are scattered in several rules, do you want a set of test cases centralized in, say, `tests/rules/prop-types.js`?

Reply:
it'd probably be ideal for every rule that uses component detection to have all of these cases, tbh